### PR TITLE
crimson: add CEPH_OSD_OP_ASSERT_VER

### PR DESCRIPTION
**Pull request by liu-chunmei**
Signed-off-by: chunmei-liu <chunmei.liu@intel.com>

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: src/crimson/osd/ops_executer.cc [hunk 1]
```diff
@@ -634,6 +634,16 @@ OpsExecuter::execute_op(OSDOp& osd_op)
     return do_read_op([this, &osd_op] (auto&, const auto& os) {
       return do_op_notify_ack(osd_op, os);
     });
+  case CEPH_OSD_OP_ASSERT_VER:
+    ++num_read;
```
Comment (rzarzynski) [inline src/crimson/osd/ops_executer.cc:638]: I propose to follow the convention and use `do_read_op()`.

File: src/crimson/osd/ops_executer.cc [hunk 2]
```diff
@@ -634,6 +634,16 @@ OpsExecuter::execute_op(OSDOp& osd_op)
...
+    if (!osd_op.op.assert_ver.ver) {
+      return crimson::ct_error::invarg::make();
+    } else if (osd_op.op.assert_ver.ver < obc->obs.oi.user_version) {
+      return crimson::ct_error::erange::make();
+    } else if (osd_op.op.assert_ver.ver > obc->obs.oi.user_version) {
+      return crimson::ct_error::value_too_large::make();
+    }
```
[the remark this answers is not in the document]
Comment (liu-chunmei) [reply src/crimson/osd/ops_executer.cc]: done.

File: src/crimson/osd/ops_executer.cc [hunk 3]
```diff
@@ -459,6 +459,19 @@ auto OpsExecuter::do_write_op(Func&& f, bool um) {
...
+  const ObjectState& os)
+{
+ if (!osd_op.op.assert_ver.ver) {
+      return crimson::ct_error::invarg::make();
+    } else if (osd_op.op.assert_ver.ver < os.oi.user_version) {
+      return crimson::ct_error::erange::make();
+    } else if (osd_op.op.assert_ver.ver > os.oi.user_version) {
+      return crimson::ct_error::value_too_large::make();
+    }
+    return seastar::now();
```
Comment (tchaikov) [inline src/crimson/osd/ops_executer.cc:473]: wrong indent.

Comment (liu-chunmei) [reply src/crimson/osd/ops_executer.cc]: done.

File: src/crimson/osd/ops_executer.cc [hunk 4]
```diff
@@ -459,6 +459,19 @@ auto OpsExecuter::do_write_op(Func&& f, bool um) {
...
+OpsExecuter::call_errorator::future<> OpsExecuter::do_assert_ver(
+  OSDOp& osd_op,
+  const ObjectState& os)
+{
+  if (!osd_op.op.assert_ver.ver) {
+    return crimson::ct_error::invarg::make();
+  } else if (osd_op.op.assert_ver.ver < os.oi.user_version) {
+    return crimson::ct_error::erange::make();
+  } else if (osd_op.op.assert_ver.ver > os.oi.user_version) {
+    return crimson::ct_error::value_too_large::make();
```
Comment (tchaikov) [inline src/crimson/osd/ops_executer.cc:471]: i am wondering if we should categorize this case as a `crimson::ct_error::erange` as well.

Comment (liu-chunmei) [reply src/crimson/osd/ops_executer.cc:471]: classic osd has the 3 checks, some test cases will check each return error.

**Comment by tchaikov** [comment]
jenkins test make check

**Comment by liu-chunmei** [comment]
jenkins test make check.

**Comment by athanatos** [comment]
@liu-chunmei Try rebasing?

**Comment by liu-chunmei** [comment]
> @liu-chunmei Try rebasing?

done.

**Review by rzarzynski — APPROVED** [review]

**Pull request merged**